From 2de99ddb894e47eac59a12d7df6096df4faaf2b9 Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 20 Oct 2025 16:22:06 +0800
Subject: [PATCH 223/254] mtk: hostapd: add support to request PP bitmap to
 driver

Add the support in 2 places
1. get_link_channel_info()
2. get_channel_info()

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 src/drivers/driver.h              | 2 ++
 src/drivers/driver_nl80211.c      | 7 +++++++
 wpa_supplicant/events.c           | 1 +
 wpa_supplicant/wpa_supplicant_i.h | 1 +
 4 files changed, 11 insertions(+)

diff --git a/src/drivers/driver.h b/src/drivers/driver.h
index ff99eb323..dc5fa6366 100644
--- a/src/drivers/driver.h
+++ b/src/drivers/driver.h
@@ -2867,6 +2867,7 @@ struct wpa_channel_info {
 	int sec_channel;
 	int center_frq1;
 	int center_frq2;
+	u16 punct_bitmap;
 	u8 seg1_idx;
 };
 
@@ -3248,6 +3249,7 @@ struct driver_sta_mlo_info {
 		u8 addr[ETH_ALEN];
 		u8 bssid[ETH_ALEN];
 		unsigned int freq, center_freq1, center_freq2;
+		u16 punct_bitmap;
 		struct t2lm_mapping t2lmap;
 		enum chan_width width;
 	} links[MAX_NUM_MLD_LINKS];
diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index ec7a31009..87fea99ae 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -1247,6 +1247,10 @@ static void get_link_channel_info(struct nlattr **link_data, u8 link_id,
 		if (link_data[NL80211_ATTR_CENTER_FREQ2])
 			info->links[link_id].center_freq2 =
 			      nla_get_u32(link_data[NL80211_ATTR_CENTER_FREQ2]);
+
+		if (link_data[NL80211_ATTR_PUNCT_BITMAP])
+			info->links[link_id].punct_bitmap =
+			      nla_get_u32(link_data[NL80211_ATTR_PUNCT_BITMAP]);
 	}
 }
 
@@ -2082,6 +2086,9 @@ static int get_channel_info(struct nl_msg *msg, void *arg)
 	if (tb[NL80211_ATTR_CENTER_FREQ2])
 		chan_info->center_frq2 =
 			nla_get_u32(tb[NL80211_ATTR_CENTER_FREQ2]);
+	if (tb[NL80211_ATTR_PUNCT_BITMAP])
+		chan_info->punct_bitmap =
+			nla_get_u32(tb[NL80211_ATTR_PUNCT_BITMAP]);
 
 	if (chan_info->center_frq2) {
 		u8 seg1_idx = 0;
diff --git a/wpa_supplicant/events.c b/wpa_supplicant/events.c
index 44de74a20..6ff11b633 100644
--- a/wpa_supplicant/events.c
+++ b/wpa_supplicant/events.c
@@ -4247,6 +4247,7 @@ static int wpa_drv_get_mlo_info(struct wpa_supplicant *wpa_s)
 		wpa_s->links[i].center_freq1 = mlo.links[i].center_freq1;
 		wpa_s->links[i].center_freq2 = mlo.links[i].center_freq2;
 		wpa_s->links[i].width = mlo.links[i].width;
+		wpa_s->links[i].punct_bitmap = mlo.links[i].punct_bitmap;
 		wpa_supplicant_update_link_bss(wpa_s, i, mlo.links[i].bssid);
 	}
 
diff --git a/wpa_supplicant/wpa_supplicant_i.h b/wpa_supplicant/wpa_supplicant_i.h
index d722f4e1f..a6eb71b42 100644
--- a/wpa_supplicant/wpa_supplicant_i.h
+++ b/wpa_supplicant/wpa_supplicant_i.h
@@ -743,6 +743,7 @@ struct wpa_supplicant {
 		u8 bssid[ETH_ALEN];
 		unsigned int freq, center_freq1, center_freq2;
 		enum chan_width width;
+		u16 punct_bitmap;
 		struct wpa_bss *bss;
 		bool disabled;
 		struct wpabuf *ies;
-- 
2.45.2

